### PR TITLE
Make postgres pod compliant with security profiles (tls-scanner)

### PR DESCRIPTION
**Pull request by rauhersu**
## Summary

Make the PostgreSQL pod compliant with the cluster's TLS security profile
and the `tls-scanner` security assessment tool, completing the TLS profile
inheritance work started in https://github.com/openshift-kni/oran-o2ims/pull/2693.

Previously, the `postgres-server` pod used static TLS defaults from the
base PostgreSQL image and was the only component in the `oran-o2ims`
namespace that did not dynamically inherit the cluster's TLS profile.
The `tls-scanner` reported it as non-compliant for both cipher suites
and PQC (Post-Quantum Cryptography) readiness.

## Backport ?

Yes, this TLS feature was present in 4.22 too.

## Changes

**PostgreSQL 18 upgrade** (`manager.yaml`, `pin_images.in.yaml`)
- Upgrade from PostgreSQL 16 to PostgreSQL 18 (`rhel9/postgresql-18:9.8`)
  which introduces `ssl_tls13_ciphers` and `ssl_groups` parameters ([PostgreSQL 18 doc](https://www.postgresql.org/docs/current/runtime-config-connection.html#RUNTIME-CONFIG-CONNECTION-SSL)) giving
  full native control over TLS 1.3 cipher suites and key exchange groups
  without filesystem-level crypto-policy overrides.

**TLS profile conversion utilities** (`tls_profile.go`)
- `TLSVersionToPostgres()` — maps OpenShift TLS version strings
  (e.g. `VersionTLS13`) to PostgreSQL's `ssl_min_protocol_version` format
  (e.g. `TLSv1.3`)
- `TLSCiphersToPostgres()` — filters TLS 1.3 cipher suites from the
  profile's cipher list, since PostgreSQL's `ssl_ciphers` only accepts
  TLS 1.2 cipher names (passing TLS 1.3 names causes a fatal startup
  error). TLS 1.3 ciphers are controlled separately via PostgreSQL 18's
  `ssl_tls13_ciphers` parameter instead.

**Dynamic PostgreSQL configuration** (`inventory_controller_postgres.go`)
- Replace static ConfigMap creation with dynamic `postgresql.conf`
  generation that appends TLS settings from the cluster profile:
  - `ssl_min_protocol_version` — enforces the cluster's minimum TLS version
  - `ssl_ciphers` — restricts TLS 1.2 ciphers to those in the profile
    (omitted under Modern profile where only TLS 1.3 is used)
  - `ssl_tls13_ciphers` — statically excludes `TLS_AES_128_CCM_SHA256`
    (not in OpenShift's approved cipher list)
  - `ssl_groups` — enables ML-KEM (`X25519MLKEM768`) for PQC key exchange
- Add `ocloud.openshift.io/tls-profile-hash` annotation to the PostgreSQL
  Deployment to trigger rolling restarts on profile changes

**Tests** (`tls_profile_test.go`, `inventory_controller_test.go`)
- Unit tests for `TLSVersionToPostgres` and `TLSCiphersToPostgres`
  covering all TLS versions, Intermediate/Modern profiles, and edge cases
- Integration test exercising three reconciliation phases: base (no TLS),
  Intermediate (mixed TLS 1.2+1.3 ciphers), and Modern (TLS 1.3 only),
  verifying ConfigMap content and annotation propagation

**Documentation** (`docs/dev/tls-configuration.md`)
- Architecture diagram showing the reconciler's dynamic `postgresql.conf`
  generation path
- `tls-scanner` invocation with `STARTTLS_PORTS=postgres=5432` for
  PostgreSQL's in-band TLS handshake
- Scanner results table showing full compliance (MinVersion, Cipher,
  PQC/ML-KEM) across all endpoints including `postgres-server`
- Notes on custom profile constraints (OpenShift rejects TLS 1.3 cipher
  names in custom `tlsSecurityProfile`)

## Test Plan

- [x] Unit tests pass (`make test`)
- [x] Envtest tests pass (`make test-envtest`)
- [x] `tls-scanner` reports full compliance under Intermediate (default) profile
- [x] `tls-scanner` reports full compliance under Modern + StrictAllComponents
- [x] Manual `openssl s_client` probes confirm TLS 1.2 rejection under Modern
- [x] Profile change triggers rolling restart of `postgres-server` pod
- [x] STARTTLS negotiation works for PostgreSQL port 5432

### Scanner Results (Modern profile, OCP 4.22 + StrictAllComponents)

All TLS endpoints pass compliance with TLS 1.3 only, ML-KEM PQC readiness,
and correct cipher suites:

| Pod | Port | TLS | PQC | API Compliance |
|-----|------|-----|-----|----------------|
| artifacts-server | 8443 | TLSv1.3 | ML-KEM | true |
| cluster-server | 8443 | TLSv1.3 | ML-KEM | true |
| hardwaremanager-server | 8443 | TLSv1.3 | ML-KEM | true |
| controller-manager | 6443 | TLSv1.3 | ML-KEM | true |
| controller-manager | 9443 | TLSv1.3 | ML-KEM | true |
| provisioning-server | 8443 | TLSv1.3 | ML-KEM | true |
| resource-server | 8443 | TLSv1.3 | ML-KEM | true |
| postgres-server | 5432 | TLSv1.3 | ML-KEM | true |

### CODE REVIEW

**Comment by rauhersu** [comment]
Not ready for review, need to tidy this up a little bit

/hold

File: docs/dev/tls-configuration.md [hunk 1]
```diff
@@ -374,9 +394,16 @@ export KUBECONFIG=/path/to/kubeconfig
...
 export NAMESPACE_FILTER="oran-o2ims"
+export STARTTLS_PORTS="postgres=5432"
 ./deploy.sh --verbose deploy
 ```
 
+The `STARTTLS_PORTS` variable enables the scanner's
+[STARTTLS support](https://github.com/openshift/tls-scanner#command-line-options)
+for PostgreSQL. Without it, the scanner cannot negotiate PostgreSQL's in-band
+TLS handshake (`SSLRequest` wire protocol) and will report `NO_TLS` for port
+5432.
```
[the remark this answers is not in the document]
Comment (rauhersu) [reply docs/dev/tls-configuration.md:436]: Rejected.

The current text is accurate and sufficient. Our doc text is deliberately concise and focused on the operator user's perspective: "set this env var or Postgres will show NO_TLS". That's the actionable takeaway.

File: docs/dev/tls-configuration.md [hunk 2]
```diff
@@ -391,31 +418,35 @@ After the scan completes, clean up cluster resources:
...
-- **All endpoints pass API Cipher Compliance** — only ciphers allowed by the
-  Intermediate profile are offered.
-- **All endpoints are PQC-ready** — ML-KEM (`X25519MLKEM768`) key exchange is
-  supported, provided by Go 1.25's native TLS 1.3 implementation.
-- **Forward Secrecy confirmed** on all TLS-enabled endpoints (only ECDHE
-  and TLS 1.3 AEAD ciphers offered).
+- **All endpoints pass API MinVersion and Cipher Compliance** — the offered
+  minimum TLS version and cipher suites match the cluster profile requirements.
+- **Go-based endpoints** achieve compliance via Go's native TLS implementation.
+  ML-KEM (`X25519MLKEM768`) support comes from Go 1.25's TLS 1.3 stack.
```
[the remark this answers is not in the document]
Comment (rauhersu) [reply docs/dev/tls-configuration.md]: Fixed.

**Comment by rauhersu** [comment]
Ready for review.

/unhold

File: docs/dev/tls-configuration.md [hunk 3]
```diff
@@ -211,16 +233,7 @@ Run a test pod inside the cluster:
...
-     echo \"=== TLS 1.3 (expect PASS) ===\"
-     echo | openssl s_client -connect resource-server.oran-o2ims.svc.cluster.local:8443 -tls1_3 2>&1 | grep -E \"Protocol|Cipher\" | head -3"],
-    "securityContext":{"allowPrivilegeEscalation":false,"capabilities":{"drop":["ALL"]}}}]}}'
+  --overrides='{"spec":{"securityContext":{"runAsNonRoot":true,"seccompProfile":{"type":"RuntimeDefault"}},"containers":[{"name":"tls-test","image":"registry.access.redhat.com/ubi9/ubi:latest","command":["bash","-c","echo \"=== TLS 1.1 (expect FAIL) ===\"; echo | openssl s_client -connect resource-server.oran-o2ims.svc.cluster.local:8443 -tls1_1 2>&1 | grep -E \"error|Protocol|Cipher\"; echo \"=== TLS 1.2 (expect PASS) ===\"; echo | openssl s_client -connect resource-server.oran-o2ims.svc.cluster.local:8443 -tls1_2 2>&1 | grep -E \"Protocol|Cipher\" | head -3; echo \"=== TLS 1.3 (expect PASS) ===\"; echo | openssl s_client -connect resource-server.oran-o2ims.svc.cluster.local:8443 -tls1_3 2>&1 | grep -E \"Protocol|Cipher\" | head -3"],"securityContext":{"allowPrivilegeEscalation":false,"capabilities":{"drop":["ALL"]}}}]}}'
```
Comment (donpenney) [inline docs/dev/tls-configuration.md]: nit: The previous multi-line format for this command is much cleaner for the reader

Comment (rauhersu) [reply docs/dev/tls-configuration.md]: Fixed, added a multi-line user friendly format.

File: internal/controllers/utils/tls_profile.go [hunk 4]
```diff
@@ -171,6 +171,44 @@ func NewTLSProfileFromEnv() configv1.TLSProfileSpec {
...
+	}
+	slog.Warn("Unrecognized TLS version for PostgreSQL, falling back to TLSv1.2",
+		slog.String("version", version))
+	return "TLSv1.2"
+}
+
+// TLSCiphersToPostgres converts a cipher list from the cluster TLS profile into
+// the colon-separated format expected by PostgreSQL's ssl_ciphers setting.
+//
+// TLS 1.3 cipher suites (prefixed with "TLS_") are excluded because PostgreSQL 16's
```
Comment (alegacy) [inline internal/controllers/utils/tls_profile.go]: [Comment authored by Claude]

nit: This says "PostgreSQL 16's ssl_ciphers" but the PR upgrades to PostgreSQL 18. The behavior is the same across versions — `ssl_ciphers` only accepts TLS 1.2 names in both PG16 and PG18 — so the version reference is just stale.

Suggestion: say "PostgreSQL's ssl_ciphers" (version-agnostic) since this is a general PostgreSQL behavior, not version-specific.

File: internal/controllers/inventory_controller_postgres.go [hunk 5]
```diff
@@ -320,11 +322,36 @@ func (t *reconcilerTask) createDatabase(ctx context.Context) (err error) {
...
+				slog.String("sslCiphers", pgCiphers))
+		}
+	}
+	// PG18 native TLS 1.3 controls:
+	// - ssl_tls13_ciphers: exclude TLS_AES_128_CCM_SHA256 (not in the OpenShift
+	//   Modern profile's allowed list, causes scanner compliance failures)
+	// - ssl_groups: enable ML-KEM (X25519MLKEM768) for PQC key exchange
+	//   alongside classical fallbacks
+	pgConf += "\nssl_tls13_ciphers = 'TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256'"
+	pgConf += "\nssl_groups = 'X25519MLKEM768:X25519:prime256v1'"
```
Comment (alegacy) [inline internal/controllers/inventory_controller_postgres.go]: [Comment authored by Claude]

suggestion: These cipher and group lists are security-sensitive configuration values inlined as string literals. Consider extracting them as named constants (e.g., in `constants/` or at the top of this file). This would make them easier to find, review, and update if OpenShift's approved list changes, and makes them greppable across the codebase.

Not a blocker — the values are correct and well-documented by the surrounding comment.

File: internal/controllers/utils/tls_profile.go [hunk 6]
```diff
@@ -171,6 +171,44 @@ func NewTLSProfileFromEnv() configv1.TLSProfileSpec {
...
+	pgVersions := map[string]string{
+		string(configv1.VersionTLS10): "TLSv1",
+		string(configv1.VersionTLS11): "TLSv1.1",
+		string(configv1.VersionTLS12): "TLSv1.2",
+		string(configv1.VersionTLS13): "TLSv1.3",
+	}
+	if v, ok := pgVersions[version]; ok {
+		return v
+	}
+	slog.Warn("Unrecognized TLS version for PostgreSQL, falling back to TLSv1.2",
```
Comment (alegacy) [inline internal/controllers/utils/tls_profile.go:187]: [Comment authored by Claude]

nit: Per project logging conventions, context-aware slog calls (`slog.WarnContext(ctx, ...)`) are preferred whenever a `ctx` is in scope. This function doesn't take a `ctx` parameter, so `slog.Warn` is technically fine — but it's worth noting that adding a `ctx` parameter would align with the convention and keep `sloglint` happy if the rule is tightened later.

Not a blocker.

File: internal/controllers/inventory_controller_test.go [hunk 7]
```diff
@@ -26,6 +26,25 @@ Test Cases in this file:
...
    - Validates complete inventory service deployment
+
+3. "NetworkPolicy scopes: API servers allow ingress-controller, database does not"
+   - Verifies API server and hardware manager NetworkPolicies allow ingress-controller,
+     monitoring, and same-namespace traffic
+   - Verifies alarms-server additionally allows alertmanager ingress
+   - Verifies database NetworkPolicy blocks ingress-controller and monitoring but
+     allows same-namespace traffic
+
+4. "Postgres ConfigMap contains TLS profile settings and deployment has tls-profile-hash"
```
Comment (alegacy) [inline internal/controllers/inventory_controller_test.go:37]: [Comment authored by Claude]

nit: Item 3 (NetworkPolicy test) was already the third `Entry` in the `DescribeTable` before this PR, but its comment block entry is being added here alongside the new item 4. This is fine — just confirming this is intentional cleanup of the previously-undocumented test case, not a misattribution.

**Review by alegacy — COMMENTED** [review]
[Comment authored by Claude]

Good PR — upgrading to PG18 native TLS controls is a clean approach that eliminates the filesystem crypto-policy override. Four inline comments below, none are blockers.

File: internal/controllers/utils/tls_profile.go (hunk 4 above)
Comment (rauhersu) [reply internal/controllers/utils/tls_profile.go]: [Comment authored by Claude]

Fixed — updated the comment to say "PostgreSQL's ssl_ciphers" (version-agnostic).

File: internal/controllers/inventory_controller_postgres.go (hunk 5 above)
Comment (rauhersu) [reply internal/controllers/inventory_controller_postgres.go]: [Comment authored by Claude]

Good call — extracted both values as named constants (`PostgresTLS13Ciphers` and `PostgresSSLGroups`) in `tls_profile.go`, and updated `inventory_controller_postgres.go` to reference them via `ctlrutils`.

File: internal/controllers/utils/tls_profile.go (hunk 6 above)
Comment (rauhersu) [reply internal/controllers/utils/tls_profile.go:187]: [Comment authored by Claude]

Acknowledged — `slog.Warn` is correct here since the function doesn't take a `ctx`. Adding one just for logging would change the signature for a speculative future linter tightening. Leaving as-is.

File: internal/controllers/inventory_controller_test.go (hunk 7 above)
Comment (rauhersu) [reply internal/controllers/inventory_controller_test.go:37]: [Comment authored by Claude]

Confirmed — intentional cleanup. The NetworkPolicy test (item 3) existed before this PR but was missing from the header comment block. Added it here alongside the new item 4 for completeness.

**Review by alegacy — APPROVED** [review]
[Comment authored by Claude]

All four review items addressed. The PG16 version reference is corrected, TLS constants are extracted with good documentation, and the controller code is cleaner for it. LGTM.

/lgtm
